server/app/auth/ApiAuthenticator.java

```diff
@@ -74,18 +78,20 @@ public void validate(Credentials rawCredentials, WebContext context, SessionStor
     // Cache the API key for quick lookup in the controller, also for subsequent requests.
     // We intentionally cache the empty optional rather than throwing here so that subsequent
     // requests with the invalid key do not put pressure on the database.
-    ApiKey apiKey =
-        apiKeyService
-            .get()
-            .findByKeyIdWithCache(keyId)
-            .orElseThrow(() -> new BadCredentialsException("API key does not exist"));
+    Optional<ApiKey> maybeApiKey = apiKeyService.get().findByKeyIdWithCache(keyId);
+
+    if (!maybeApiKey.isPresent()) {
+      unauthorized(context, "API key does not exist: " + keyId);
```

Comment:
I believe this is causing the unit test failure from CI since it's doing an exact match on the error message.

Reply:
There's a few different pieces:

* API key: the database-backed resource that provides auth semantics for API calls
* API key ID: a random string that uniquely identifies an API key
* API key secret: a random string that, paired with the ID, allows an API caller to use an API key
* API key credentials: the combination of API key ID and secret, joined on ':' and base-64 encoded

We log the key ID here so that maintainers of a deployment can see which keys callers are attempting and failing to authenticate with. We never log the secret.

Reply:
ah woops, yeah i meant to reply to your other comments. sorry!